Rewrite HTTP server definitions section

lmolkova opened the pull request:
Fixes #414

## Changes

- rewrites HTTP server definitions section: removes unused terminology and focuses on HTTP framework instrumentation experience
- adds example diagrams
- Removes repetitive notes from yaml replacing them with links

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [x] [CHANGELOG.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CHANGELOG.md) updated for non-trivial changes.
* [ ] ~~schema-next.yaml~~

trask commented on docs/http/http-spans.md:
```diff
@@ -295,52 +298,56 @@ See the examples for more details about:
+HTTP server instrumentations SHOULD do the best effort when populating `server.address` and `server.port` attributes and determine them by using the first of the following that applies:
+
+* Explicit user configuration (if supported).
+* Server name and port available through web framework or HTTP server configuration (if known to the instrumentation).
```
wondering if this sounds like you should use [ServletRequest#getServerName()](https://docs.oracle.com/javaee/7/api/javax/servlet/ServletRequest.html#getServerName--) and [ServletRequest#getServerName()](https://docs.oracle.com/javaee/7/api/javax/servlet/ServletRequest.html#getServerPort--) (which I think isn't what we want)

lmolkova replied:
Agree. I thought more about it and

- removed explicit config and added a remark that applications can overwrite anything with processors.
- gave the highest priority to the `Forwarded` and friends as they'd match what the client has set
- clarified that framework/web-server info should come from users (and not from `Host` header)

thoughts?

trask commented on docs/http/http-spans.md:
```diff
@@ -295,52 +298,56 @@ See the examples for more details about:
+* The original host which may be passed by the reverse proxy in the `host` identifier of the [`Forwarded`][Forwarded] header or in the [`X-Forwarded-Host`][X-Forwarded-Host] header.
+* The `Host` header.
+
+As a result, `server.*` attributes capture logical host and port name and intended to be the same on the client and server sides. Application developers MAY capture private host information using applicable [resource attributes](/docs/resource/README.md).
```
this might be nice up at the top instead, as it provides the motivation for our definition

trask reviewed (commented):
I think this is very helpful

trask commented on docs/http/http-spans.md:
```diff
@@ -295,52 +298,58 @@ See the examples for more details about:
+* The `Host` header.
+
+> **Note**: The `Host` header specifies host and port number of the server. The same applies to the `host` identifier of `Forwarded` header, the `X-Forwarded-Host` header, or the authority component in the request target. Instrumentations SHOULD populate both `server.address` and `server.port` attributes by parsing the applicable property.
+
+The `server.address` SHOULD NOT be set if only IP address is available and capturing name would require a reverse DNS lookup.
```
#413

```suggestion
```

trask commented on docs/http/http-spans.md:
```diff
@@ -295,52 +298,58 @@ See the examples for more details about:
 ## HTTP server
 
-To understand the attributes defined in this section, it is helpful to read the "Definitions" subsection.
+Read the following section to understand how HTTP server instrumentations capture server information.
```
```suggestion
Read the following section to understand how HTTP server instrumentations should capture server information.
```

jsuereth replied:
should is normative and must be capitalized, FYI

lmolkova replied:
this part is informational and non-normative. There is a normative section ("Setting `server.address` and `server.port` attributes") where normative verbs are capitalized.

jsuereth replied:
I know this is awkward, but we should avoid any normative verb in this section of docs.

trask commented on docs/http/http-spans.md:
```diff
@@ -295,52 +298,58 @@ See the examples for more details about:
--- SRV.9 in [JSR 53][];
-in a deployment of a Python application to Apache, the application would be the [PEP 3333][] conformant callable that is configured using the
-[`WSGIScriptAlias` directive][modwsgisetup] of `mod_wsgi`).
+HTTP server frameworks and their instrumentations have a limited knowledge about the HTTP infrastructure and intermediaries request went through. In a general case, they can only use HTTP request properties such as request target or headers to populate `server.*` attributes.
```
```suggestion
HTTP server frameworks and their instrumentations have limited knowledge about the HTTP infrastructure and intermediaries that requests go through. In a general case, they can only use HTTP request properties such as request target or headers to populate `server.*` attributes.
```

trask commented on docs/http/http-spans.md:
```diff
@@ -295,52 +298,58 @@ See the examples for more details about:
+* The original host which may be passed by the reverse proxy in the `host` identifier of the [`Forwarded`][Forwarded] header or in the [`X-Forwarded-Host`][X-Forwarded-Host] header.
+* Server name and port available through web framework or HTTP server properties if they're based on the user configuration. Instrumentations SHOULD NOT use properties that rely on the `Host` header.
+* The [request target][request target] if it's passed in the absolute form.
+* The `Host` header.
```
Fallback for HTTP/1.0 where no `Host` header? (`network.local.address/port`?)

lmolkova replied:
Let's figure out `server.address` requirement level first (#424).
If it stays recommended, then I'd prefer to not provide a fallback option.

HTTP 1.0 is [obsolete since 1997](https://en.wikipedia.org/wiki/HTTP#:~:text=Summary%20of%20HTTP%20milestone%20versions) and it would be nice not to add features for it specifically.

trask replied:
> Let's figure out `server.address` requirement level first (#424). If it stays recommended, then I'd prefer to not provide a fallback option.

I'm good with keeping `server.address` as (only) recommended, and sent #429 with your proposal for fixing the requirement level on `server.port` instead.

I'm not sure (yet) about not having a fallback.

We've discussed the idea that reconstructing the original URL on the server side is an important use case.

I think users seeing something like `http://123.123.123.123:8080/petclinic` is potentially more useful than not seeing anything.

lmolkova replied:
If address is recommended, then backends should expect that it might be missing. By keeping it recommended, we're also assuming that URL reconstruction is not essential. But it can still be done by falling back (on the consumer) to network attributes.

It also feels like something that can be changed post-stability without breaking based on real-world feedback.

lmolkova commented:
After discussion with @trask:
- removed request-target in absolute form from options. Based on the [RFC7230](https://datatracker.ietf.org/doc/html/rfc7230#section-5.4) "A client MUST send a Host header field in an HTTP/1.1 request even if the request-target is in the absolute-form", so there is no need in mentioning it
- removed framework/web-server options. The motivation:
  - the server does not know its domain name (e.g. when `foo.io` is handled by `wus3-foo.io`, the latter one does not need to know the original DNS name the client reached).
  - if there is some configuration, this is an HTTP proxy/lb configuration and would not match the original DNS name


Also, added `:authority` HTTP/2 pseudo-header which is interchangeable with HTTP1.1 Host and may be set without the Host header. If both are set, they SHOULD match.

jsuereth commented on docs/http/http-spans.md:
```diff
@@ -294,52 +291,49 @@ See the examples for more details about:
 ## HTTP server
 
-To understand the attributes defined in this section, it is helpful to read the "Definitions" subsection.
+Read the following section to understand how HTTP server instrumentations should capture server information.
```
If this isn't intended to be normative, please remove `should` for any other word.

e.g.

`Read the following section to understand how HTTP server instrumentations are suggested to capture server information.`

joaopgrassi commented on model/metrics/http.yaml:
```diff
@@ -6,36 +6,18 @@ groups:
+          See [Setting `server.address` and `server.port` attributes](/docs/http/http-spans.md#setting-serveraddress-and-serverport-attributes).
 
-          Warning: since this attribute may be based on HTTP headers, opting in to it may allow an attacker
+          > **Warning**: since this attribute may be based on HTTP headers, opting in to it may allow an attacker
```
If you want the Warning here to contain the nice icon, this has to be on its own like, like

> **Warning**
> since this attribute..

lmolkova replied:
cool, thanks for the pointer!

joaopgrassi commented on docs/http/http-spans.md:
```diff
@@ -294,52 +291,47 @@ See the examples for more details about:
+For example, different versions of the same web-application can run side-by-side as independent applications behind the reverse proxy which routes request to one or another based on the request path.
+
+Instances of different HTTP server applications may run on the same physical host and share the same IP address, but listen to different TCP/UDP ports.
+In order to route request to a specific application, reverse proxies usually modify the [HTTP Host header][Host and authority] replacing the original value provided by the client with an actual proxied server name. This behavior depends on the reverse proxy configuration. In some cases, `Host` header is not used when routing request to a specific application, making it prone to having bogus conte ...
```
```suggestion
In order to route the request to a specific application, reverse proxies usually modify the [HTTP Host header][Host and authority] replacing the original value provided by the client with an actual proxied server name. This behavior depends on the reverse proxy configuration. In some cases, the `Host` header is not used when routing a request to a specific application, making it prone to having bogus content.
```

joaopgrassi commented on docs/http/http-spans.md:
```diff
@@ -294,52 +291,47 @@ See the examples for more details about:
+
+#### Setting `server.address` and `server.port` attributes
+
+In the context of HTTP server, `server.address` and `server.port` attributes capture the original host name and port. The are intended, whenever possible, to be the same on the client and server sides.
```
```suggestion
In the context of HTTP server, `server.address` and `server.port` attributes capture the original host name and port. They are intended, whenever possible, to be the same on the client and server sides.
```
